Rewrite of documentation

vickycl opened the pull request:
## Purpose of this PR

- Extracts import, export and 'USD in the Editor' into separate pages
- Adds missing details in several areas
- Changed some images and removed duplicates
- Removed a lot of detail from the Readme and point to the Documentation entry point instead, to reduce maintenance burden

## Testing

**Functional Testing status:** N/A

<!-- Explanation of what's tested, how tested and existing or new automation tests. Can include manual testing by self and/or QA. Specify test plans. Rarely acceptable to have no testing.  -->

**Performance Testing status:** N/A

<!-- Could this PR affect performance? If so, what has been done to measure time taken / memory used etc? Can include new or existing performance tests. Also see [Ensuring Performance by Default](https://confluence.unity3d.com/display/DEV/Ensuring+Performance+by+Default). -->

## Overall Product Risks
<!-- See Testing Status, Complexity and Halo Effect for your PR](https://confluence.unity3d.com/display/DEV/Testing+Status%2C+Complexity+and+Halo+Effect+for+your+PR). -->

**Complexity:** 0
<!-- (Minimal / Low / Medium / High) -->

**Halo Effect:** 0
<!-- (Minimal / Low / Medium / High) -->

## Additional information

**Note to reviewers:**

<!-- Info per person for what to focus on, or historical info to understand who have previously reviewed and coverage. Help them get context. -->

**Reminder:**
<!-- Things to remember to do before finalizing this PR. -->
- [ ] Add entry in CHANGELOG.md _(if applicable)_

vickycl committed:
Initial rewrite of documentation

- extract import, export and USD in the Editor into separate pages
- add missing details in several areas
- changed some images and removed duplicates
- removed a lot of detail from the Readme and point to the Documentation entry point instead, to reduce maintenance burden

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/Images/usd-import-options.PNG:
I just realized i have never used "Unload Subtree" and "Load Subtree" and it seems like it unloads / loads all payloads within an imported USD object

I feel like we should either
1. Explain what these buttons do (and maybe even rename them to be consistent with Load All / Unload All payloads?)
2. Or delete them from the menu since we already have a place for this function in the USD Asset component

vickycl replied:
The only benefit I can guess at of this over the UsdPayload tickbox is that you don't have to manually find all of your payloads to load them as long as they are below a common ancestor. Theoretically if you wanted to load all of the payloads under an 'environment' parent but not the ones under a 'character' parent without finding each individual payload (and that parent isn't itself a payload) then you could use this. You're right, it should be documented, and possibly have 'Payload' in the name for clarity.

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD Export.md:
```diff
@@ -0,0 +1,77 @@
+
+Note: Modifications to the transform of the Root GameObject are not currently reflected in the override, as Unity assumes the root in all USD files is at the origin.
+
+## Export Selected as USDZ
```
we should note that Unity GameObjects set with "EditorOnly" tag will be ignored for export

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD Export.md:
```diff
@@ -0,0 +1,77 @@
+# Exporting USD files
```
we should note that Unity GameObjects set with "EditorOnly" tag will be ignored for export

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
-* API Compatibility Level .NET 4.x is no longer required but still provides better performances than .NET 2   
-   In Edit > Project Settings > Player :   
-    ![USD .NET version](Images/USD_.NET_version.png)
+* API Compatibility Level .NET 4.x is no longer required but still provides better performances than .NET 2 
```
This doesn't read as a requirement - Could simplify the wording as  'Recommendation' or add a section for that.

vickycl replied:
It actually isn't an option in later versions of Unity that we support, so to avoid confusion I'll remove it entirely :)

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD Export.md:
```diff
@@ -0,0 +1,77 @@
+
+## Export Selected with Children
+
+On Mac you will see multiple file format options- USDC and USDA.
```
we could probably add a space between the "options" and "-"
though now im curious if what you did is the correct form in the UK

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
 
-The USD importer works with linear color space only. To ensure colors are imported correctly,
-set the project color space to "linear" in Edit > Project Settings > Player:
+The USD importer works with linear color space only. To ensure colors are imported correctly, set the project color space to "linear" in Edit > Project Settings > Player:
```
Sounds like a Requirement, could be added to the relevant section.

vickycl replied:
Good point - I've added a project configuration section for this and a few other notes that aren't strict requirements but worth noting.

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
-
->  **Note:** This feature has no dependency to and is not based on the Recorder package.
-
+* Importing and exporting USD:
```
Maintenance - the list of features is duplicated in two places. (Here &  top-level readme.) Makes sense to have this for the repo & package cases though.

 Without a preprocessor tool (Mustache or similar) we need to make sure to manually replicate both of these so good job doing so here. For interest, splitting to a separate  features 'include' file is also not viable as Security concerns has always blocked this feature (see https://github.com/github/markup/issues/172#issuecomment-33241601)

vickycl replied:
Yeah, I have already updated the USD.md and not the Readme.md at least once 😂 I didn't know about include, that's interesting! What a shame we can't use it 😞

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
-   In Edit > Project Settings > Player :   
-    ![USD .NET version](Images/USD_.NET_version.png)
+* API Compatibility Level .NET 4.x is no longer required but still provides better performances than .NET 2 
+In Edit > Project Settings > Player : ![USD .NET version](Images/USD_.NET_version.png)
```
Seems like you deleted 
Images/USD_.NET_version.png

in this PR

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
+        * Load All and Individual
+    * Layer stacks
+* Geometry:
+    * UV Set
```
I kind of want to specify somewhere in the README that viewing UV channels within Unity is available on Unity 2020+
Just like as a reminder

What do you think?

vickycl replied:
That's not a specification of this package but of USD, so I don't think this is the right place :)

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
+    * Camera
+        * We do not currently import and export all physical camera settings
+    * Materials:
+        * Transparency Settings (these can be set manually)
```
Transparency & Double-Sided Settings

vickycl replied:
Just in case there's other things we're missing that we don't even know about, I'll keep this vague and put them as examples?

lee-aandrew replied:
yeah that sounds good since its going to be near impossible to test all types of materials and their flags

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
-set the project color space to "linear" in Edit > Project Settings > Player:
+The USD importer works with linear color space only. To ensure colors are imported correctly, set the project color space to "linear" in Edit > Project Settings > Player:
 
 ![USD linear](Images/USD_linear.png)
```
"Images/USD_linear.png" has been deleted in this PR

vickycl replied:
All of the images in the `Images` folder were duplicated in `package/com.unity.formats.usd/Documentation~/Images`, so I've only removed the duplicates at the higher level. This path is relative so *should* use the remaining one.

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
+
+Once the USD package is installed, a USD menu will appear, enabling you to easily import and export USD files.
 
 ![USD menu](Images/USD_menu.png)
```
"USD_menu.png" has been deleted in this PR

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
-settings under Edit > Project Settings > Quality:
+In Unity 2019, the USD importer supports importing unlimited weights per vertex. However, to see the effect of more than 4 weights per vertex, this must be enabled in the project settings under Edit > Project Settings > Quality:
 
 ![USD unlimited_weights](Images/USD_unlimited_weights.png)
```
"Images/USD_unlimited_weights.png" has been deleted in this pr

vickycl replied:
As above, this points to the one in the documentation folder.

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
-
+* Importing and exporting USD:
+    * Import as GameObject, Prefab, or Timeline Clip
+        * Formats: USDA, USDC, USDZ
```
"USD" format is omitted here 😄

vickycl replied:
I was going for the encoding format rather than the file endings here... a .usd file just defaults to USDC encoding I believe

vickycl replied:
Is it clearer now?

lee-aandrew replied:
yes, thank you for the change!

lee-aandrew commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
+    * Import as GameObject, Prefab, or Timeline Clip
+        * Formats: USDA, USDC, USDZ
+    * Export Game Objects to USD
+        * Formats: USDA, USDC, USDZ
```
"USD" format is omitted here as well

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD Import.md:
```diff
@@ -0,0 +1,53 @@
+
+## Importing Options
+
+-   **Import As GameObjects**: Imports selected USD file as GameObjects directly into the Scene Hierarchy.
```
Would be good to elaborate on the benefits of the GameObjects vs Prefab option? As a 'new user' I'm really not clear which option should be my go-to option to start with. Will I find a limitation from one vs the other etc.

vickycl replied:
I've added some clarification, does this help?

CraigHutchinson replied:
Definitely helps my understanding as it really helps to have some explanation of the difference.

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD Import.md:
```diff
@@ -0,0 +1,53 @@
+- Converting from USD-supported mixed face meshes to a triangular mesh.
+
+Change vertex winding order: 
+- Unity uses a clockwise mesh winding order, where other DCCs may use anti-clockwise.
```
Localization: American English would prefer `counterclockwise` (https://books.google.com/ngrams/graph?content=counterclockwise,anticlockwise&year_start=1900&year_end=2000&corpus=17&smoothing=3&share=) while British speakers are indifferent - I hear Unity docs prefer US standardization?

vickycl replied:
Ack. Didn't even clock that (geddit?) Thanks for pointing it out!

lee-aandrew replied:
TIL the colonizers use the word "anticlockwise"

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD Export.md:
```diff
@@ -0,0 +1,77 @@
+
+Modifications to transforms in USD files can be exported as overrides to the original files. 
+
+Overrides can be exported from the USD menu with the GameObject containing the [UsdAsset](/USD-in-the-Editor#USD-Asset) component selected, which will export transform overrides for the entire hierarchy. Alternatively, you can also export just the overrides when exporting from the Recorder window by changing the 'Override Setting' to 'Export Transform Overrides Only'.
```
Dead-link: `/USD-in-the-Editor#USD-Asset` is does not work from Github (https://github.com/Unity-Technologies/usd-unity-sdk/blob/documentation/rewrite/package/com.unity.formats.usd/Documentation~/USD%20Export.md)

I think removing the leading '/' (root folder) shall resolve this as the file is in the same folder.

vickycl replied:
Thanks!

lee-aandrew reviewed (changes requested):
Added some comments - but awesome changes! 🏆

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD Export.md:
```diff
@@ -0,0 +1,77 @@
+
+## Export Transform Overrides
+
+Modifications to transforms in USD files can be exported as overrides to the original files. 
```
Sounds like a limitation for 'known limitations', I'm not sure? Is this maybe this is just the more 'general' limitation that Unity and USD have been developed with differing goals I guess.

vickycl replied:
I'm not sure what you mean by this being a limitation? Exporting the Transform Overrides is a different import option to plain export, exporting only the USD 'Overs' for a USD file instead of the full file. It's a USD composition concept to allow layering. Does this answer your question?

CraigHutchinson replied:
This is an odd one... this comment was on line 48 ... confused how this happened!

> Note: Modifications to the transform of the Root GameObject are not currently reflected in the override, as Unity assumes the root in all USD files is at the origin.

vickycl replied:
Oh! Yes in that case I completely understand what you mean by Unity and USD being developed differently. Yes I guess it is a limitation in that case that the Exporter always sets the root to the origin. This feels a lot more specific than the limitations currently called out in that section, kind of like this is more of an implementation detail? If I was looking for the reason why my overrides were being exported at the root I'd probably look for it in this specific section, rather than a broad limitations section. What do you think?

CraigHutchinson replied:
I definitely think you are right. I'm still learning our stack so being a bit nit-picky. I'd be fine with sorting this out in a future revision, as a 'user' isn't necessarily going to read through all the documentation. This could be addressed as FAQ and/or exporter/importer specific issues section at a later date.

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD.md:
```diff
@@ -1,126 +1,114 @@
 ## Requirements
 
 * Windows / OSX (Intel only)
 * Unity version: 2019.4 and up
```
Sanity-check: Is there no upper-limit on supported Unity version?

lee-aandrew replied:
I haven't done a full testing on our trunk, but yes the package works on latest unity

CraigHutchinson reviewed (changes requested):
One necessary change:
- There is an internal broken link to `/USD-in-the-Editor#USD-Asset`

Other than that looks good, made observations which may be applied in future work as desired 👍

vickycl committed:
Initial fix ups from PR reviews

vickycl committed:
Update README.md

vickycl committed:
Fix ups from PR.

- Moved some requirements around for clarity
- Fixed some internal links (hopefully)
- Removed some defunct settings and their corresponding images

vickycl committed:
Include a mention of the USD version in use.

vickycl committed:
Merge branch 'dev' into documentation/rewrite

vickycl committed:
Replaced image with one showing the new Subtree menu items.

CraigHutchinson commented on package/com.unity.formats.usd/Documentation~/USD Export.md:
```diff
@@ -0,0 +1,77 @@
+
+Modifications to transforms in USD files can be exported as overrides to the original files. 
+
+Overrides can be exported from the USD menu with the GameObject containing the [UsdAsset](package/com.unity.formats.usd/Documentation~/USD-in-the-Editor.md#USD-Asset) component selected, which will export transform overrides for the entire hierarchy. Alternatively, you can also export just the overrides when exporting from the Recorder window by changing the 'Override Setting' to 'Export Transfor ...
```
Dead-link: I think this needs to use relative paths? When this is linked from Github it takes you too https://github.com/Unity-Technologies/usd-unity-sdk/blob/documentation/rewrite/package/com.unity.formats.usd/Documentation~/package/com.unity.formats.usd/Documentation~/USD-in-the-Editor.md#USD-Asset

while it should be https://github.com/Unity-Technologies/usd-unity-sdk/blob/documentation/rewrite/package/com.unity.formats.usd/Documentation~/USD-in-the-Editor.md
That's without the duplication on `package/com.unity.formats.usd/Documentation~/` 

Maybe this was required so it works on the package though and we can't have both github+package links working? I'd otherwise expected `USD-in-the-Editor.md#USD-Asset` could have worked.

vickycl replied:
Thank you for testing this again! I think I've fixed it now as it works for me, is it fixed for you?

CraigHutchinson replied:
Yes, I just tested it from GitHub and looks to work nicely now :)

vickycl committed:
Fix dead link

CraigHutchinson reviewed (approved):
Looks good now

lee-aandrew reviewed (approved):
LGTM! 👍

michaeljblain commented on README.md:
```diff
@@ -1,242 +1,83 @@
+    * Vertex Colour
+* Instancing:
+    * Point Instancing
+    * Scene Instancing
```
Change this to Scenegraph instancing to link with USD docs.

michaeljblain commented on package/com.unity.formats.usd/Documentation~/USD Import.md:
```diff
@@ -0,0 +1,53 @@
+
+## Importing Materials
+
+To import materials from USD, firstly import the USD file using the USD menu. Then, to get the materials, change the 'Import Settings' > 'Materials' in the Inspector to 'Import Preview Surface'. Finally re-import the USD file.
```
We can start this paragraph in the same way as "Importing Payloads" i.e. "Materials are not loaded by default".

michaeljblain commented on package/com.unity.formats.usd/Documentation~/USD Import.md:
```diff
@@ -0,0 +1,53 @@
+- As above, meshes also go through a basis conversion as required.
+
+Triangulate meshes:
+- Converting from USD-supported mixed face meshes to a triangular mesh.
```
Maybe change "mixed face" to "arbitrary polygon"

michaeljblain commented on package/com.unity.formats.usd/Documentation~/USD-in-the-Editor.md:
```diff
@@ -0,0 +1,94 @@
+* **Don’t Load Payloads (default)** - Import without any payloads.
+* **Load All** - Import and load all payloads.
+ 
+ Payloads can also be individually loaded and unloaded on the individual objects.
```
Change "individual" to "specific" to remove duplicate use of word.

michaeljblain replied:
change "objects" to "prims"

michaeljblain reviewed (changes requested):
A great update to the docs. Some minor suggestions.

vickycl committed:
Updates from PR review.

vickycl committed:
Merge branch 'dev' into documentation/rewrite

vickycl committed:
Fixed links?

vickycl committed:
Merge branch 'documentation/rewrite' of https://github.com/Unity-Technologies/usd-unity-sdk into documentation/rewrite

vickycl committed:
Fixed missing image?

vickycl committed:
Fixed image.. again
